subsys/net/lib/dns/dns_sd.c

```diff
@@ -0,0 +1,786 @@
+			  uint8_t *buf, uint16_t buf_offset, uint16_t buf_size,
+			  uint16_t *service_offset, uint16_t *instance_offset,
+			  uint16_t *domain_offset);
+static int add_txt_record(const struct dns_sd_rec *inst, uint32_t ttl,
+			  uint16_t instance_offset, uint8_t *buf,
+			  uint16_t buf_offset, uint16_t buf_size);
+static int add_aaaa_record(const struct dns_sd_rec *inst, uint32_t ttl,
+			   uint16_t host_offset, const uint8_t addr[16],
+			   uint8_t *buf, uint16_t buf_offset,
+			   uint16_t buf_size);
+static int add_srv_record(const struct dns_sd_rec *inst, uint32_t ttl,
+			  uint16_t instance_offset, uint16_t domain_offset,
+			  uint8_t *buf, uint16_t buf_offset, uint16_t buf_size,
+			  uint16_t *host_offset);
+static bool dns_sd_rec_is_valid(const struct dns_sd_rec *inst);
+
+#endif
+
+/**
+ * Calculate the size of a DNS-SD service
+ *
+ * This macro calculates the size of the DNS-SD service for a DNS
+ * Resource Record (RR).
+ *
+ * For example, if there is a service called 'My Foo'._http._tcp.local.,
+ * then the returned size is 18. That is broken down as shown below.
+ *
+ * - 1 byte for the size of "_http"
+ * - 5 bytes for the value of "_http"
+ * - 1 byte for the size of "_tcp"
+ * - 4 bytes for the value of "_tcp"
+ * - 1 byte for the size of "local"
+ * - 5 bytes for the value of "local"
+ * - 1 byte for the trailing NUL terminator '\0'
+ *
+ * @param ref the DNS-SD record
+ * @return the size of the DNS-SD service for a DNS Resource Record
+ */
+size_t dns_sd_service_proto_size(const struct dns_sd_rec *ref)
+{
+	return 0
+	       + DNS_LABEL_LEN_SIZE + strlen(ref->service)
+	       + DNS_LABEL_LEN_SIZE + strlen(ref->proto)
+	       + DNS_LABEL_LEN_SIZE + strlen(ref->domain)
+	       + DNS_LABEL_LEN_SIZE
+	;
+}
+
+/**
+ * Check Label Validity according to RFC 1035, Section 3.5
+ *
+ * <label> ::= <letter> [ [ <ldh-str> ] <let-dig> ]
+ * <ldh-str> ::= <let-dig-hyp> | <let-dig-hyp> <ldh-str>
+ * <let-dig-hyp> ::= <let-dig> | -
+ * <let-dig> ::= <letter> | <digit>
+ * <letter> ::= [a-zA-Z]
+ * <digit> ::= [0-9]
+ */
+bool label_is_valid(const char *label, size_t label_size)
+{
+	size_t i;
+
+	if (label == NULL) {
+		return false;
+	}
+
+	if (label_size == 0) {
+		/* automatically calculate the length of the string */
+		label_size = strlen(label);
+	}
+
+	if (label_size < DNS_LABEL_MIN_SIZE ||
+	    label_size > DNS_LABEL_MAX_SIZE) {
+		return false;
+	}
+
+	for (i = 0; i < label_size; ++i) {
+		if (isalpha((int)label[i])) {
+			continue;
+		}
+
+		if (i > 0) {
+			if (isdigit((int)label[i])) {
+				continue;
+			}
+
+			if ('-' == label[i]) {
+				continue;
+			}
+		}
+
+		return false;
+	}
+
+	return true;
+}
+
+
+static bool instance_is_valid(const char *instance)
+{
+	size_t i;
+	size_t instance_size;
+
+	if (instance == NULL) {
+		LOG_DBG("label is NULL");
+		return false;
+	}
+
+	instance_size = strlen(instance);
+	if (instance_size < DNS_SD_INSTANCE_MIN_SIZE) {
+		LOG_DBG("label '%s' is too small (%zu, min: %u)",
```

Comment:
Please use the networking debug macros in these networking files, so `NET_DBG()` here